### ISSUE TITLE
`mpfr::frexp()` overload removed may cause compilation errors

### ISSUE THREAD

**Issue by vdilecce** [advanpix/mpreal#26]
I was looking at version 3.7.0, and I noticed that overload `frexp(const mpreal&, int*)` was removed due to being ambiguous.

I am guessing that the ambiguity may be that `mpfr_exp_t` could be different types, and if it was `int` there would be two identical overloads (by the way, why using `mpfr_exp_t` in `mpfr::frexp()` and `mp_exp_t` everywhere else?)

However, I see a possible problem in simply removing such overload. I mostly have encountered scenarios with the exponent type of GMP/MPFR being `long*`, whereas `std::frexp` uses `int*`. Consider the following templated code designed to work with different types (e.g. `double` and `mpfr::mpreal`):

```cpp
template<typename T>
void myFunction()
{
    T   num {12.3};
    int exp {};
    T   var {frexp(num, &exp)}; // Argument-dependent lookup
}
```

By removing overload `frexp(const mpreal&, int*)` the above code does not compile (pointer conversion mismatch). The following example, which avoids ADL, shows the same problem:

```cpp
namespace mp
{
    template<typename T> const T            frexp(const T            num, int* exp) { return  std::frexp(num, exp); }
    template<> inline    const mpfr::mpreal frexp(const mpfr::mpreal num, int* exp) { return mpfr::frexp(num, exp); }
}

template<typename T>
void myFunction(T x)
{
    T   num {12.3};
    int exp {};
    T   var {mp::frexp(num, &exp)};
}
```

Is there a way to restore the removed `frexp(const mpreal&, int*)` overload? Maybe in one of the following ways:

- Conditional compilation (GMP):

    ```cpp
    #if __GMP_MP_SIZE_T_INT != 1
        // Define mpfr::frexp(const mpreal&, mp_exp_t*) overload
    #endif
    ```

- Conditional compilation (MPFR):

    ```cpp
    #if _MPFR_EXP_FORMAT != 2
        // Define mpfr::frexp(const mpreal&, mpfr_exp_t*) overload
    #endif
    ```

- Templating `mpfr::frexp()` (which would probably require `std::is_same::value`)

**Comment by advanpix** [comment]
Valid points. The issue is caused by the mess with integer types in GMP & MPFR and compatibility with older versions.
Previously MPFR exponent used to be of `mp_exp_t` type (<3.0.0). The `mpfr_exp_t` was introduced later on (and these types might not be the same anymore).

The `_MPFR_EXP_FORMAT` macro defines the actual type, but it is an internal macro and it is better to avoid it....

I agree, `mpreal` should have overload compatible with C++ standard:
`inline const mpreal frexp(const mpreal& x, int* exp, mp_rnd_t mode = mpreal::get_default_rnd())`

Will do the changes shortly. Thank you very much for suggestion.

**Comment by advanpix** [comment]
Fixed by  6ad3016f33329ba8f5b6b4bb956e275abe6da411
